include/dqlite.h

```diff
@@ -17,8 +17,304 @@
+ * connections originating from the same process.
+ */
+int dqlite_server_set_bind_address(dqlite_server *server, const char *addr);
+
+/**
+ * Configure the function that this server will use to connect to other servers.
+ *
+ * The same function will be used by the server's attached client to establish
+ * connections to all servers in the cluster. @arg is a user data parameter that
+ * will be passed to all invocations of the connect function.
+ */
+int dqlite_server_set_connect_func(dqlite_server *server,
+				   dqlite_connect_func f,
+				   void *arg);
+
+/**
+ * Configure the target number of voting nodes for this server's cluster.
+ *
+ * When this server is the cluster leader, it will try to promote and demote
+ * other servers to reach the target voter count. You should configure the same
+ * target on all servers that will participate in the cluster.
+ */
+int dqlite_server_set_target_voters(dqlite_server *server, int n);
+
+/**
+ * Configure the target number of standby nodes for this server's cluster.
+ *
+ * When this server is the cluster leader, it will try to promote and demote
+ * other servers to reach the target standby count. You should configure the
+ * same target on all servers that will participate in the cluster.
+ */
+int dqlite_server_set_target_standbys(dqlite_server *server, int n);
+
+/**
+ * Start running the server.
+ *
+ * Once this function returns successfully, the server will be ready to accept
+ * client requests using the functions below.
+ */
+int dqlite_server_start(dqlite_server *server);
+
+/**
+ * Hand over the server's privileges to other servers.
+ *
+ * This is intended to be called before dqlite_server_stop. The server will try
+ * to surrender leadership and voting rights to other nodes in the cluster, if
+ * applicable. This avoids some disruptions that can result when a privileged
+ * server stops suddenly.
+ */
+int dqlite_server_handover(dqlite_server *server);
+
+/**
+ * Stop the server.
+ *
+ * The server will stop processing requests from client or other servers. To
+ * smooth over some possible disruptions to the cluster, call
+ * dqlite_server_handover before this function. After this function returns
+ * (successfully or not), you should call dqlite_server_destroy to free
+ * resources owned by the server.
+ */
+int dqlite_server_stop(dqlite_server *server);
+
+/**
+ * Free resources owned by the server.
+ *
+ * This should be called after dqlite_server_stop.
+ */
+void dqlite_server_destroy(dqlite_server *server);
+
+/**
+ * Open a database connection on the dqlite cluster.
+ *
+ * This request will be transparently forwarded to the cluster leader as needed.
+ *
+ * This is the analogue of sqlite3_open. @name is the name of the database
+ * to open, which will be created if it does not exist. All servers in the
+ * cluster share a "namespace" for databases.
+ */
+int dqlite_open(dqlite_server *server, const char *name, dqlite **db);
```

Comment:
One idea to make `dqlite_open` a bit more future-proof (I'm thinking to the goal of supporting creating clients without requiring a locally running server, for example for implementing interactive remote shells) would be to change this signature to something like:

```C
int dqlite_open(void *connector, const char *name, dqlite **db, int flags);
```

where the meaning of `connector` would be defined using a flag.

For example:

```C
dqlite_server *server;
dqlite_server_create(..., &server);
...
dqlite_open(server, "my-db", DQLITE_OPEN_USING_SERVER);
```

or similar names. In the future we could also support passing something that is not a server instead, but rather some other object that just knows about the addresses of some servers in the cluster and that can connect to them.